test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,387 @@
+		Config: &runtime.ContainerConfig{
+			Metadata: &runtime.ContainerMetadata{
+				Name: "",
+			},
+			Image: &runtime.ImageSpec{
+				Image: imageWindowsNanoserver,
+			},
+			// Command for cloned containers
+			Command: []string{
+				"cmd",
+				"/c",
+				"ping",
+				"-t",
+				"127.0.0.1",
+			},
+			Annotations: map[string]string{
+				"io.microsoft.virtualmachine.templateid": "",
+			},
+		},
+	}
+}
+
+func waitForTemplateSave(ctx context.Context, t *testing.T, templatePodID string) {
+	app := "hcsdiag"
+	arg0 := "list"
+	for {
+		cmd := exec.Command(app, arg0)
+		stdout, err := cmd.Output()
+		if err != nil {
+			t.Fatalf("failed while waiting for save template to finish: %s", err)
+		}
+		if strings.Contains(string(stdout), templatePodID) && strings.Contains(string(stdout), "SavedAsTemplate") {
+			break
+		}
+		timer := time.NewTimer(time.Millisecond * 100)
+		select {
+		case <-ctx.Done():
+			t.Fatalf("Timelimit exceeded for wait for template saving to finish")
+		case <-timer.C:
+		}
+	}
+}
+
+func createPodAndContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, sandboxRequest *runtime.RunPodSandboxRequest, containerRequest *runtime.CreateContainerRequest) (podID, containerID string) {
+	podID = runPodSandbox(t, client, ctx, sandboxRequest)
+	containerRequest.PodSandboxId = podID
+	containerRequest.SandboxConfig = sandboxRequest.Config
+	containerID = createContainer(t, client, ctx, containerRequest)
+	startContainer(t, client, ctx, containerID)
+	return podID, containerID
+}
+
+// Creates a template sandbox and then a template container inside it.
+// Since, template container can take time to finish the init process and then exit (at which
+// point it will actually be saved as a template) this function wait until the template is
+// actually saved.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createTemplateContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templateSandboxRequest *runtime.RunPodSandboxRequest, templateContainerRequest *runtime.CreateContainerRequest) (templatePodID, templateContainerID string) {
+	templatePodID, templateContainerID = createPodAndContainer(ctx, t, client, templateSandboxRequest, templateContainerRequest)
+
+	// Send a context with deadline for waitForTemplateSave function
+	d := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(ctx, d)
+	defer cancel()
+	waitForTemplateSave(ctx, t, templatePodID)
+	return
+}
+
+// Creates a clone from the given template pod and container.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createClonedContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templatePodID, templateContainerID string, cloneNumber int) (clonedPodID, clonedContainerID string) {
+
+	cloneSandboxRequest := getClonedPodConfig()
```

Comment:
Commented about this above but can we not just pass in an index and the templatePodID into getClonedPodConfig/getClonedContainerConfig and set them there instead of doing this setup in the caller multiple times?